### PR TITLE
Fix Windows

### LINKED ISSUE #81: Windows build failing due to missing include header
Build on MSVC fails due to missing decl for memcpy. Line 94 at src/wasm-config.h.in,

```
#elif COMPILER_IS_MSVC

#define WASM_UNUSED
#define WASM_WARN_UNUSED _Check_return_
#define WASM_INLINE __inline
#define WASM_STATIC_ASSERT(x) _STATIC_ASSERT(x)
#define WASM_UNLIKELY(x) (x)
#define WASM_LIKELY(x) (x)
#define WASM_PRINTF_FORMAT(format_arg, first_arg)

__inline unsigned long wasm_clz_u32(unsigned long mask) {
  unsigned long index;
  _BitScanReverse(&index, mask);
  return sizeof(unsigned long) * 8 - (index + 1);
}

__inline unsigned long wasm_clz_u64(unsigned __int64 mask) {
#if _M_X64
  unsigned long index;
  _BitScanReverse64(&index, mask);
  return sizeof(unsigned __int64) * 8 - (index + 1);
#elif _M_IX86
  unsigned long index;
  unsigned long high_mask;
  memcpy(&high_mask, (unsigned char*)&mask + sizeof(unsigned long),
         sizeof(unsigned long));
  if (_BitScanReverse(&index, high_mask)) {
    return sizeof(unsigned long) * 8 - (index + 1);
  }

  unsigned long low_mask;
  memcpy(&low_mask, &mask, sizeof(unsigned long));
  _BitScanReverse(&index, low_mask);
  return sizeof(unsigned __int64) * 8 - (index + 1);
#else
#error unexpected architecture
#endif
}
```

Issue is fixed by `include <string.h>`

### PR DESCRIPTION
**Pull request by Cellule**
Fixes #81 

Several forward & backward slash fixes where paths were assumed to always have forward slash.

Changed stdout/stderr output to binary to eliminate \r\r\n outputs

Changed the lexer to use binary files and teach it about carriage return when searching for line offset. 
-fseek() on an offset taken from a text file in mode text is unreliable because there can be some conversion on fread() (e.g.: \r\n => \n)

Changed the diff tool to remove/ignore end lines, there were too many differences between the output on Windows and Linux with end lines.

### CODE REVIEW

File: src/sexpr-wasm.c [hunk 1]
```diff
@@ -301,6 +301,17 @@ static void on_module_begin(uint32_t index, void* user_data) {
       get_module_filename(ctx->allocator, &ctx->output_filename_noext, index);
   if (index != 0)
     wasm_writef(&ctx->json_stream, ",\n");
+  const int l = strlen(ctx->module_filename);
+  char* filename = ctx->module_filename;
+  int i = 0;
```
Comment (binji) [inline src/sexpr-wasm.c]: make this into a function?

File: src/sexpr-wasm.c [hunk 2]
```diff
@@ -301,6 +301,17 @@ static void on_module_begin(uint32_t index, void* user_data) {
...
+  for (; i < l; ++i)
+  {
```
Comment (binji) [inline src/sexpr-wasm.c]: nit: braces on previous line

File: src/sexpr-wasm.c [hunk 3]
```diff
@@ -301,6 +301,17 @@ static void on_module_begin(uint32_t index, void* user_data) {
...
+      if (filename[i] == '\\')
```
Comment (binji) [inline src/sexpr-wasm.c]: nit: 2 space indent

File: src/wasm-ast-lexer.c [hunk 4]
```diff
@@ -515,6 +515,7 @@ static WasmResult scan_forward_for_line_offset_in_buffer(
   int line = buffer_line;
   int line_offset = 0;
   const char* p;
+  size_t is_previous_carriage = 0;
```
Comment (binji) [inline src/wasm-ast-lexer.c]: Use `WasmBool` instead

File: src/wasm-common.h [hunk 5]
```diff
@@ -371,6 +371,8 @@ void wasm_default_binary_error_callback(uint32_t offset,
                                         const char* error,
                                         void* user_data);
 
+void wasm_initialize_stdio();
```
Comment (binji) [inline src/wasm-common.h]: nit: `wasm_init_stdio` is more consistent with other function names

File: src/wasm-interp-sq.c [hunk 6]
```diff
@@ -122,7 +122,7 @@ static void on_option(struct WasmOptionParser* parser,
       break;
 
     case FLAG_HELP:
-      wasm_print_help(parser);
+      wasm_print_help(parser, "wasm-interp-sq");
```
Comment (binji) [inline src/wasm-interp-sq.c]: How about a `#define` at the top of the file for this:
```
#define PROGRAM_NAME "wasm-interp-sq"
```

File: test/run-tests.py [hunk 7]
```diff
@@ -117,10 +117,10 @@ def Indent(s, spaces):
 
 
 def DiffLines(expected, actual):
-  expected_lines = expected.splitlines(1)
-  actual_lines = actual.splitlines(1)
+  expected_lines = [line for line in expected.splitlines() if len(line) > 0]
```
Comment (binji) [inline test/run-tests.py]: nit: `if line` seems more Pythonic to me

**Comment by binji** [comment]
The Travis build seems to be failing, any idea what's broken?

**Comment by Cellule** [comment]
There seems to be a problem with the json encoding, I'll verify.

**Comment by Cellule** [comment]
It doesn't repro on my Linux box, can you verify if it repros for you ?

File: src/sexpr-wasm.c [hunk 8]
```diff
@@ -334,8 +346,14 @@ static void on_command(uint32_t index,
   Context* ctx = user_data;
   if (index != 0)
     wasm_writef(&ctx->json_stream, ",\n");
+  const int l = strlen(loc->filename);
+  char* filename = wasm_alloc(ctx->allocator, l, WASM_DEFAULT_ALIGN);
```
Comment (binji) [inline src/sexpr-wasm.c]: + 1 for null terminator.

**Comment by binji** [comment]
Yeah, it looks like it is happening because the filename isn't null-terminated.

**Comment by Cellule** [comment]
Thanks, I forgot that strlen() didn't count the null terminator

File: src/sexpr-wasm.c [hunk 9]
```diff
@@ -301,6 +312,8 @@ static void on_module_begin(uint32_t index, void* user_data) {
       get_module_filename(ctx->allocator, &ctx->output_filename_noext, index);
   if (index != 0)
     wasm_writef(&ctx->json_stream, ",\n");
+  ctx->module_filename = convert_backslash_to_forward_slash(ctx->module_filename, strlen(ctx->module_filename));
```
Comment (binji) [inline src/sexpr-wasm.c]: nit: wrap line to 80 columns

File: src/wasm-common.c [hunk 10]
```diff
@@ -150,3 +155,15 @@ void wasm_default_binary_error_callback(uint32_t offset,
   else
     fprintf(stderr, "error: @0x%08x: %s\n", offset, error);
 }
+
+void wasm_init_stdio()
+{
```
Comment (binji) [inline src/wasm-common.c]: nit: { on previous line

File: src/wasm-common.c [hunk 11]
```diff
@@ -150,3 +155,15 @@ void wasm_default_binary_error_callback(uint32_t offset,
...
+#if COMPILER_IS_MSVC
+    int result = _setmode(_fileno(stdout), _O_BINARY);
```
Comment (binji) [inline src/wasm-common.c]: nit: 2 space indent

**Comment by binji** [comment]
lgtm w/ a few nits

**Comment by binji** [comment]
Sorry, forgot to merge this.

**Pull request merged**